### ISSUE TITLE
Empty source file with hyperlinked source

### ISSUE THREAD

**Issue by alexbiehl** [haskell/haddock#730]
@harpocrates Alec, Herbert generated and uploaded the documentation with haddock master for unix-2.7.2.2. 

Unfortunately https://hackage.haskell.org/package/unix-2.7.2.2/docs/src/System.Posix.Files.Common.html#unionFileModes looks pretty empty! Could you have a look? It might be related to your recent lexer changes.

He also regenerated the documentation for lens-4.15.4 and looking at its hyperlinked source everything seems fine https://hackage.haskell.org/package/lens-4.15.4/docs/src/Control.Lens.At.html#at

**Comment by hvr** [comment]
I've just tried to build it again via `new-haddock`, and btw, `--haddock-hyperlink-source` would try to use HsColour, so instead I used `--haddock-options="--hyperlinked-source --quickjump"`. However, this currently segfaults for me... I need to investigate a bit more.

However, for building the `unix` docs iirc I used an older `cabal sandbox` based script of mine, which does basically something like

```bash
$CABAL sandbox init
$CABAL install --dep -j4 --enable-documentation
$CABAL configure -O0
$CABAL haddock --with-haddock=$HADDOCK --for-hackage --haddock-option=--hyperlinked-source --haddock-option=--quickjump
```

**Comment by harpocrates** [comment]
@hvr I knew something was up with `cabal haddock`'s `--hyperlinked-source`! Using `--haddock-options="--hyperlinked-source"`  on `unix` hangs for me. I'll investigate.

**Comment by hvr** [comment]
@harpocrates does it really *hang* infinitely, or does just take a several minutes to complete? :)

**Comment by harpocrates** [comment]
@hvr I mean _hang_. I've found the problem. Technically, it is GHC, through `addSourceToTokens` that is hanging. The problem is probably the `LINE` pragmas. Chasing the issue into GHC now.

```haskell
-- Causes a hang on my machine, perhaps other things on other machines
module Main where

{-# LINE 4 "Dummy" #-}

-- | A @'main'@ multi
main :: IO ()
main = pure ()
```

**Comment by hvr** [comment]
Yes, the `LINE` pragma seems indeed to be cause for producing empty syntax-highlighted html, in my case, running `haddock --hyperlinked-source --quickjump  Main.hs` resulted in the followng `src/Main.html`:

```html
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd"><html xmlns="http://www.w3.org/1999/xhtml"><head><link rel="stylesheet" type="text/css" href="style.css" /><script type="text/javascript" src="highlight.js"></script></head><body><pre><span class="hs-comment">-- Causes a hang on my machine, perhaps other things on other machines</span><span>
</span><a name="line-2"></a><span class="hs-keyword">module</span><span> </span><span class="hs-identifier">Main</span><span> </span><span class="hs-keyword">where</span><span>
</span><a name="line-3"></a><span>
</span><a name="line-4"></a><span>
</span><a name="line-5"></a><span>
</span><a name="line-6"></a><span> </span><span> </span><span> </span><span>
</span><a name="line-7"></a><span> </span><span> </span><span> </span><span>
</span><a name="line-8"></a></pre></body></html>
```

**Comment by harpocrates** [comment]
I've nailed the problem. A fix is on the way.

@hvr we end up calling `uncons` on an empty string buffer, so I'm not surprised that my machine hangs while yours segfaults.

**Comment by harpocrates** [comment]
OK. The fix is not easy. In a nutshell: when GHC's lexer encounters a `LINE` or `COLUMN` pragma, it doesn't emit _anything_ - it just updates its internal state. That is annoying because there is no way of observing this in the tokens outputted. The modified locations cause GHC problems in `addSourceToTokens`. I see two possible avenues:

  1. modify GHC's lexer to add an option which forces it to emit actual tokens (note that there is an `ITline_pragma` token used nowhere in GHC's codebase, but there is no token constructor for the column pragma)
  2. have Haddock deal with these pragmas (sort of how it handles CPP)

I'm implementing 2 right now. I think it is less elegant, but I'm not really interested in trying to prod yet another change through GHC.

**Comment by hvr** [comment]
I'm still confused why `uncons` would ever (unsafely) segfault or hang, rather than throw an ordinary (safe) Haskell exception (which could be caught).

**Comment by harpocrates** [comment]
@hvr  The problem is that `addSourceToTokens` can get caught in a loop, continuously `uncons`ing a `StringBuffer` (trying to reach a certain location which doesn't exist) long past the time when it has meaningful data.

**Comment by harpocrates** [comment]
I've opened https://phabricator.haskell.org/D4336 to support lexing `LINE`/`COLUMN` pragmas as actual tokens. If that gets merged by, there should be no Haddock-side change needed. If it cannot be merged in time for 8.4.1, I think we should move the GHC-lexer changes off to the next release.

**Comment by harpocrates** [comment]
I've just checked this is fixed in GHC 8.4.1 via https://phabricator.haskell.org/D4336 and https://github.com/haskell/haddock/commit/ac33472e834d381f95fd56586e57e6653263055c.

The annoyance in https://github.com/haskell/haddock/issues/730#issuecomment-357310773 is fixed in https://github.com/haskell/haddock/commit/97b0189927924b82ee26f762c88ccd965eee8d80.
